sessionctx/variable/varsutil.go

```diff
@@ -137,6 +137,28 @@ func SetSessionSystemVar(vars *SessionVars, name string, value types.Datum) erro
 	return vars.SetSystemVar(name, sVal)
 }
 
+// ValidateSetSystemVar checks if system variable satisfies specific restriction.
+func ValidateSetSystemVar(name string, value string) (v string, warn error, err error) {
+	if name == GroupConcatMaxLen {
+		var ivalue uint64
+		ivalue, err = strconv.ParseUint(value, 10, 64)
+		if err != nil {
+			return "", nil, ErrWrongTypeForVar.GenByArgs(GroupConcatMaxLen)
+		}
+		// The reasonable range of 'group_concat_max_len' is 4~18446744073709551615(64-bit platforms)
+		// See https://dev.mysql.com/doc/refman/8.0/en/server-system-variables.html#sysvar_group_concat_max_len for details
+		if ivalue < 4 {
+			warn = ErrTruncatedWrongValue.GenByArgs(GroupConcatMaxLen, strconv.FormatUint(ivalue, 10))
+			ivalue = 4
+		} else if ivalue > 18446744073709551615 {
```

Comment:
in mysql:
```sql
MySQL(localhost:3306) > set @@group_concat_max_len=18446744073709551616;
ERROR 1232 (42000): Incorrect argument type to variable 'group_concat_max_len'
MySQL(localhost:3306) > set @@group_concat_max_len=1;
Query OK, 0 rows affected, 1 warning (0.00 sec)

MySQL(localhost:3306) > show warnings;
+---------+------+-----------------------------------------------------+
| Level   | Code | Message                                             |
+---------+------+-----------------------------------------------------+
| Warning | 1292 | Truncated incorrect group_concat_max_len value: '1' |
+---------+------+-----------------------------------------------------+
1 row in set (0.00 sec)

MySQL(localhost:3306) > select version();
+-----------+
| version() |
+-----------+
| 5.7.22    |
+-----------+
1 row in set (0.01 sec)
```

Reply:
@zz-jason with my patch, the result is the same with MySQL. see below: 
```sql
tidb> set @@group_concat_max_len=18446744073709551616;
ERROR 1232 (42000): Incorrect argument type to variable 'group_concat_max_len'
tidb> set @@group_concat_max_len=1;
Query OK, 0 rows affected, 1 warning (0.00 sec)

tidb> show warnings;
+---------+------+-----------------------------------------------------+
| Level   | Code | Message                                             |
+---------+------+-----------------------------------------------------+
| Warning | 1292 | Truncated incorrect group_concat_max_len value: '1' |
+---------+------+-----------------------------------------------------+
1 row in set (0.00 sec)
```
Actually the first test will fail in line 206, `varsutil.go`:
```go
val, err := strconv.ParseUint(value, 10, 64)
if err != nil {
	return value, ErrWrongTypeForVar.GenByArgs(name)
}
```
The check & reset in line 214 actually would'n happen